src/circuit.rs

```diff
@@ -50,13 +270,401 @@ impl plonk::Circuit<pallas::Base> for Circuit {
+            let psi_old = self.load_private(
+                layouter.namespace(|| "witness psi_old"),
+                config.advices[0],
+                self.psi_old,
+            )?;
+
+            // Witness rho_old
+            let rho_old = self.load_private(
+                layouter.namespace(|| "witness rho_old"),
+                config.advices[0],
+                self.rho_old.map(|rho| rho.0),
+            )?;
+
+            // Witness cm_old
+            let cm_old = Point::new(
+                config.ecc_chip(),
+                layouter.namespace(|| "cm_old"),
+                self.cm_old.as_ref().map(|cm| cm.to_affine()),
+            )?;
+
+            // Witness g_d_old
+            let g_d_old = Point::new(
+                config.ecc_chip(),
+                layouter.namespace(|| "gd_old"),
+                self.g_d_old.as_ref().map(|gd| gd.to_affine()),
+            )?;
+
+            // Witness ak.
+            let ak: Option<pallas::Point> = self.ak.as_ref().map(|ak| ak.into());
+            let ak = Point::new(
+                ecc_chip.clone(),
+                layouter.namespace(|| "ak"),
+                ak.map(|ak| ak.to_affine()),
+            )?;
+
+            // Witness nk.
+            let nk = self.load_private(
+                layouter.namespace(|| "witness nk"),
+                config.advices[0],
+                self.nk.map(|nk| *nk),
+            )?;
+
+            // Witness v_old.
+            let v_old = self.load_private(
+                layouter.namespace(|| "witness v_old"),
+                config.advices[0],
+                self.v_old
+                    .map(|v_old| pallas::Base::from_u64(v_old.inner())),
+            )?;
+
+            // Witness v_new.
+            let v_new = self.load_private(
+                layouter.namespace(|| "witness v_new"),
+                config.advices[0],
+                self.v_new
+                    .map(|v_new| pallas::Base::from_u64(v_new.inner())),
+            )?;
+
+            (rho_old, psi_old, cm_old, g_d_old, ak, nk, v_old, v_new)
+        };
+
+        // Merkle path validity check.
+        // TODO: constrain output to equal public input
+        let _anchor = {
+            let merkle_inputs = MerklePath {
+                chip_1: config.merkle_chip_1(),
+                chip_2: config.merkle_chip_2(),
+                domain: SinsemillaHashDomains::MerkleCrh,
+                leaf_pos: self.pos,
+                path: self.path,
+            };
+            let leaf = *cm_old.extract_p().inner();
+            merkle_inputs.calculate_root(layouter.namespace(|| "MerkleCRH"), leaf)?
+        };
+
+        // Value commitment integrity.
+        // TODO: constrain to equal public input cv_net
+        let _cv_net = {
+            // v_net = v_old - v_new
+            let v_net = {
+                let v_net_val = self.v_old.zip(self.v_new).map(|(v_old, v_new)| {
+                    // Do the subtraction in the scalar field.
```

Comment:
Because we know that v_old and v_new are in [0, 2<sup>64</sup>), it would be equivalent to do this in the base field, and maybe slightly simpler.

Reply:
We are now doing this in the base field.